### ISSUE TITLE
armoryctl atecc command and Error : sysfs-i2c: no such device or address

### ISSUE THREAD

**Issue by Grifenfer** [f-secure-foundry/armoryctl#3]
Hello (yeah again), when I reboot the usbarmory and run  the command **armoryctl atecc info** before the command  **armoryctl atecc self_test**, the command return the following error : _Error : sysfs-i2c: no such device or address_. (Same for new commands I add to the tool) 

Any idea why ? 

Thank you in advance.

**Comment by abarisani** [comment]
1. Ensure that i2c modules are loaded:

```
sudo lsmod |grep -i i2c

i2c_dev                16384  0
i2c_imx                20480  0
i2c_core               40960  2 i2c_imx,i2c_dev
```

2. Ensure that you are running armoryctl with sudo:

```
sudo ./armoryctl atecc info
serial:0x... revision:0x00006002
```

**Comment by Grifenfer** [comment]
1.  Same results as you.
2. I am in root mode (sudo su -)

In fact the problem comes for the first command(s) I type, regardless if it is info or self_test. 

But when it works once, it doesn't fail anymore. And after a make command in the armoryctl directory it directly works fine.

**Comment by Grifenfer** [comment]
To resume the situation : I am in root mode (sudo su -) and I have the same result as you for the lsmod |grep -i i2c command and the i2cdetect command sees the device at address 0x60. 

And what ever I am doing, the info and selftest commands return the error message what ever I am doing (i2cdetect, lsmod..., make) before I run them. And when it works fine once, it works fine after.

**Comment by abarisani** [comment]
We think this is due to the device needing a sync and/or wake-up depending on the I2C bus state, we are pondering options to mitigate it in `armoryctl` library.

FYI `i2cget` (the Linux command to query the I2C bus) does the same, though it would make sense to work this around in our library.

**Comment by Grifenfer** [comment]
I don't if it is helpful but this problem came to me after you changed the library to change the sleep state to an idle state in the ExecuteCmd function.

**Comment by abarisani** [comment]
Can you try and re-introduce [this sleep](https://github.com/f-secure-foundry/armoryctl/commit/4b7296d9ca6ea7eca959f88ef0894a0af28192d1#diff-79e7ed6f3bae6f4b8e3507cfd2b5005cf50709299a778f0c910d9165416aaaecL154) to see if it solves the issue?

It can be re-added as:

time.Sleep(250 * time.Millisecond)

**Comment by Grifenfer** [comment]
The problem seems to be solved with it  

This has to be exactely 250 ms (max execution time) or this can be the time of the execution of the command ?

**Comment by abarisani** [comment]
I am currently testing narrowing it down per datasheet, it should be:

```go
time.Sleep(1500 * time.Microseconds)
```

which corresponds to the tWHI delay.

The command execution time (not the wake up) is variable, for this reason the library always sleeps the `CmdMaxExecutionTime`, however this can be overridden for commands known to be faster with the `CmdExecutionTime` variable or, in case of quickly repeated commands that happen before 1.5 seconds with the `wake` boolean.

**Comment by abarisani** [comment]
Can you please test [this fix](https://github.com/f-secure-foundry/armoryctl/commit/174fd9e17f8f8930f91028895bc9134b24fb5778) and close the issue if it works? Thanks!

**Comment by Grifenfer** [comment]
time.Sleep(1500 * time.Microsecond) is better  (Microsecond without s).  And it works also with time.Sleep(CmdExecutionTime).

However when I shut down my usbarmory (unplug it) and replug it I don't have the response (neither an error) of the command typed (example: armoryctl atecc info). I have to use the make command to have the response again.

**Comment by abarisani** [comment]
Typo fixed in my last commit.

I am afraid I don't get the second part, can you rephrase please?

**Comment by Grifenfer** [comment]
Nevermind it works. 

I said that when I rebooted the usbarmory and ran an armoryctl command, it was not printing the resulting string (for info command, the serial number) but it works now.

Thank you for your responses and reactivity.
